src/main/java/technology/dice/dicefairlink/discovery/members/FairlinkMemberFinder.java

```diff
@@ -0,0 +1,163 @@
+  private static final Logger LOGGER = Logger.getLogger(FairlinkMemberFinder.class.getName());
+  private static final ExclusionTag EXCLUSION_TAG = new ExclusionTag("Fairlink-Exclude", "true");
+  private static final Set<String> EMPTY_SET = new HashSet<>(0);
+
+  private final FairlinkConfiguration fairlinkConfiguration;
+  private final MemberFinderMethod memberFinder;
+  private final ReplicaValidator replicaValidator;
+  private final Function<Collection<String>, SizedIterator<String>> iteratorBuilder;
+  protected final FairlinkConnectionString fairlinkConnectionString;
+  protected final TagFilter tagFilter;
+  protected Optional<String> fallbackEndpoint = Optional.empty();
+  protected Collection<String> excludedInstanceIds = new HashSet<>(0);
+
+  public FairlinkMemberFinder(
+      FairlinkConfiguration fairlinkConfiguration,
+      FairlinkConnectionString fairlinkConnectionString,
+      ScheduledExecutorService tagsPollingExecutor,
+      TagFilter excludedInstancesFinder,
+      MemberFinderMethod memberFinder,
+      Function<Collection<String>, SizedIterator<String>> stringSizedIteratorBuilder,
+      ReplicaValidator replicaValidator) {
+    this.fairlinkConnectionString = fairlinkConnectionString;
+    this.fairlinkConfiguration = fairlinkConfiguration;
+    this.tagFilter = excludedInstancesFinder;
+    this.memberFinder = memberFinder;
+    this.replicaValidator = replicaValidator;
+    final Duration startJitter = fairlinkConfiguration.randomBoundDelay();
+    this.iteratorBuilder = stringSizedIteratorBuilder;
+    LOGGER.info("Starting excluded members discovery with " + startJitter + " delay.");
+    tagsPollingExecutor.scheduleAtFixedRate(
+        () -> excludedInstanceIds = safeExclusionsDiscovery(),
+        startJitter.getSeconds(),
+        fairlinkConfiguration.getTagsPollerInterval().getSeconds(),
+        TimeUnit.SECONDS);
+  }
+
+  private Set<String> safeExclusionsDiscovery() {
+    try {
+      return tagFilter.listExcludedInstances(EXCLUSION_TAG);
+    } catch (Exception e) {
+      LOGGER.log(
+          Level.SEVERE, "Could not discover exclusions; including all discovered instances", e);
+      return EMPTY_SET;
+    }
+  }
+
+  public final SizedIterator<String> discoverReplicas() {
+    try {
+      long before = System.currentTimeMillis();
+      final ClusterInfo clusterInfo = this.memberFinder.discoverCluster();
+      this.fallbackEndpoint = Optional.of(clusterInfo.getReadonlyEndpoint());
+      final Set<String> filteredReplicas =
+          clusterInfo.getReplicas().stream()
+              .filter(db -> !excludedInstanceIds.contains(db))
+              .filter(
+                  dbIdentifier ->
+                      (!this.fairlinkConfiguration.isValidateConnection())
+                          || this.validate(fairlinkConfiguration.hostname(dbIdentifier)))
+              .map(dbIdentifier -> fairlinkConfiguration.hostname(dbIdentifier))
+              .collect(Collectors.toSet());
```

Comment:
`ImmutableSet.toImmutableSet`

Reply:
This project doesn't have a dependency on Guava (only on test scope)